gdal/doc/source/development/rfc/rfc76_ogrpythondrivers.rst

```diff
@@ -0,0 +1,397 @@
+
+    QGIS has now a way to create Python-based providers such as
+    in https://github.com/qgis/QGIS/blob/master/tests/src/python/provider_python.py
+    Having a way to do in GDAL itself also allows the rest of GDAL/OGR based
+    tools to use the OGR Python driver.
+
+How does that work ?
+--------------------
+
+Driver registration
++++++++++++++++++++
+
+The driver registration mechanism is extended to look for .py scripts in a
+dedicated directory ({prefix}/lib/gdalplugins/python if hardcoded on compilation
+for Unix builds, or the GDAL_PYTHON_DRIVER_PATH environment variable).
+Those Python script must have in their first lines at least 2 global variables
+set:
+
+- ``DRIVER_NAME = "short_name"``
+- ``DRIVER_SUPPORTED_API_VERSION = 1`` . Currently only 1 supported. If the
+  interface changed in a backward incompatible way, we would increment internally
+  the supported API version number. This item enables us to check if we are able
+  to "safely" load a Python driver. If a Python driver would support several API
+  versions (not clear if that's really possible at that point), it might use an
+  array syntax to indicate that, like ``[1,2]``
+- ``DRIVER_DCAP_VECTOR = "YES"``
+- ``DRIVER_DMD_LONGNAME = "my super plugin"``
+
+Optional metadata such as ``DRIVER_DMD_EXTENSIONS`` or
+``DRIVER_DMD_HELPTOPIC`` can be defined (basically, any driver metadata key string prefixed by
+``DRIVER_``
+
+Those will be parsed in a pure textual way, without invocation of the Python
+interpreter, both for efficiency consideration and also because we want to
+delay the research or launch of the Python interpreter as much as possible
+(the typical use case if GDAL used by QGIS: we want to make sure that QGIS
+has itself started Python, to reuse that Python interpreter)
+
+From the short metadata, the driver registration code can instantiate GDALDriver
+C++ objects. When the Identify() or Open() method is invoked on that object,
+the C++ code will:
+
+* start/find the Python interpreter if not already done
+* load the .py file as a Python module in the Python interpreter if not already
+* invoke the Python ``create_driver`` method, if not already done, that will
+  instantiate a ``Driver`` object with a ``identify`` and ``open`` method.
+* call the  ``identify`` and ``open`` method depending on the originated API call.
+
+The ``open`` method will return a Python ``Dataset`` object with required and
+optional methods that will be invoked by the corresponding GDAL API calls. And
+likewise for the ``Layer`` object. See the example_.
+
+Connection with the Python interpreter
+++++++++++++++++++++++++++++++++++++++
+
+The logic will be shared with the VRT Python fix functionality. It relies on
+runtime loading of the Python library, rather than compile time linking. The
+reason is that we do not know in advance with which Python version GDAL can
+potentially be linked, and we do not want gdal.so/gdal.dll to be explicitly linked
+with a particular Python library.
```

Comment:
This explains my earlier question in https://github.com/OSGeo/gdal/pull/1984/files#r342839730 (a footnote missing)